client/ts/plots/heatmap.ts

```diff
@@ -0,0 +1,107 @@
+import Plotly from 'plotly.js-dist';
+
+/**
+ * Creates a heatmap from raw data using plotly.
+ * @param x An array of categorical values. Number or string
+ * @param y An array of categrorical values. Numnber or string
+ * @param xName Title for the x-axis
+ * @param yName Title for the y-axis
+ * @param chartHolderId The id of the html element to render the chart in.
+ */
+export function heatmap(x:Array<any>, y:Array<any>, xName:string, yName:string, chartHolderId:string): void {
+    let data = processForHeatmap(x, y);
+    let chartData = [
+        {
+            x: data['x'],
+            y: data['y'],
+            z: data['z'],
+            type: 'heatmap',
+            hoverongaps: false
+        }
+    ];
+    let layout = {
+        xaxis: {
+            title: {
+                text: xName
+            }
+        },
+        yaxis: {
+            title: {
+                text: yName
+            }
+        }
+    };
+    Plotly.newPlot(chartHolderId, chartData, layout);
+}
+
+/**
+ * Transoform raw data into a form the heatmap function of Plotly can use. 
+ * @param x An array of categorical values. Number or string
+ * @param y An array of categrorical values. Numnber or string
+ */
+export function processForHeatmap(x:Array<any>, y:Array<any>) {
+    let x_labels = unique(x);
+    let y_labels = unique(y);
+
+    let x_label_index = getLabelIndexDict(x_labels);
+    let y_label_index = getLabelIndexDict(y_labels);
+
+    let counts = create2dZeroMatrix(x_labels.length, y_labels.length);
+
+    let count_max = -99999;
+    let count_min = 99999;
+    let n = x.length;
+    for (let i = 0; i < n; i++){
+        counts[y_label_index[y[i]]][x_label_index[x[i]]] ++
+        count_max = Math.max(counts[y_label_index[y[i]]][x_label_index[x[i]]], count_max)
```

Comment:
where are count_max and count_min used for? They seems not affecting the counts value

Reply:
It might be needed when we want to manually set the color gradient.

Reply:
I will comment them out for now.

Reply:
done